src/basicfuns.jl

```diff
@@ -153,8 +153,37 @@ Return `log(1+exp(x))` evaluated carefully for largish `x`.
 This is also called the ["softplus"](https://en.wikipedia.org/wiki/Rectifier_(neural_networks))
 transformation, being a smooth approximation to `max(0,x)`. Its inverse is [`logexpm1`](@ref).
 """
-log1pexp(x::Real) = x < 18.0 ? log1p(exp(x)) : x < 33.3 ? x + exp(-x) : oftype(exp(-x), x)
-log1pexp(x::Float32) = x < 9.0f0 ? log1p(exp(x)) : x < 16.0f0 ? x + exp(-x) : oftype(exp(-x), x)
+function log1pexp(x::Real)
+    t = log1p(exp(-abs(x)))
+    return x ≤ 0 ? t : t + x
+end
+
+function log1pexp(x::Union{Float16,Float32,Float64})
+    #= Faster implementation for floats. Based on Mächler (2012), available at:
+    https://cran.r-project.org/web/packages/Rmpfr/vignettes/log1mexp-note.pdf =#
+	a, b, c = _log1pexp_branch_bounds(x)
+    if x ≤ a
+        return exp(x)
+    elseif x ≤ b
+        return log1p(exp(x))
+    elseif x ≤ c
+        return x + exp(-x)
+    else
+        return x
+    end
+end
+
+#=
+Given the `approx` used in a branch of log1pexp(x) above, we find the first `x` (from above
+or below) that is a root of
+
+    T(log1pexp(big(x))) - approx(T(x))
+
+This determines the branch bounds below.
+=#
+_log1pexp_branch_bounds(::Float64) = (-37, 18, 33.3)
+_log1pexp_branch_bounds(::Float32) = (-15, 9, 14.5)
+_log1pexp_branch_bounds(::Float16) = (-7, 3, 5.7)
```

Comment:
Are these the same as reported in the reference? It is probably better to make them all of type Float64, Float32, and Float16 to avoid promotions in the conditional statements.

Reply:
I would have thought the compiler would take care of the promotions (isn't this some form of constant propagation?)